File: src/cleaning/page_furniture.py
import re
from collections import Counter
from src.models import DocumentPage
EDGE_LINES = 2      
REPEAT_RATIO = 0.6 
MIN_PAGES_TO_JUDGE = 3  
DIGITS = re.compile(r"\d+")
PAGE_NUMBER = re.compile(
    r"^(page\s*)?[ivxlcdm\d]+(\s*(/|of|-)\s*\d+)?$", re.IGNORECASE
)

def _fingerprint(text: str) -> str:
    """'Page 3 of 40' and 'Page 4 of 40' must look like the same line."""
    return DIGITS.sub("#", text.lower()).strip()

def find_furniture(pages: list[DocumentPage]) -> set[str]:
    if len(pages) < MIN_PAGES_TO_JUDGE:
        return set()
    counts: Counter[str] = Counter()
    for page in pages:
        edge = page.lines[:EDGE_LINES] + page.lines[-EDGE_LINES:]
        for fingerprint in {_fingerprint(line.text) for line in edge}:
            counts[fingerprint] += 1
    threshold = len(pages) * REPEAT_RATIO
    return {f for f, n in counts.items() if f and n >= threshold}

def strip_furniture(pages: list[DocumentPage]) -> list[DocumentPage]:
    furniture = find_furniture(pages)
    judgeable = len(pages) >= MIN_PAGES_TO_JUDGE
    cleaned: list[DocumentPage] = []
    for page in pages:
        last = len(page.lines) - 1
        kept = []
        for index, line in enumerate(page.lines):
            at_edge = index < EDGE_LINES or index > last - EDGE_LINES
            if _fingerprint(line.text) in furniture:
                continue
            if at_edge and judgeable and PAGE_NUMBER.match(line.text.strip()):
                continue
            kept.append(line)
        cleaned.append(
            DocumentPage(
                text = "\n".join(line.text for line in kept),
                page_number = page.page_number,
                source = page.source,
                lines = kept,
            )
        )
    return cleaned

File: src/cleaning/pageFurniture.ts
import type { DocumentPage } from "../models.ts";

const EDGE_LINES = 2; // only the top/bottom lines can be furniture
const REPEAT_RATIO = 0.6; // on 60%+ of pages -> it is a header/footer
const MIN_PAGES_TO_JUDGE = 3; // with 1-2 pages, repetition proves nothing

const DIGITS = /\d+/g;
const PAGE_NUMBER = /^(page\s*)?[ivxlcdm\d]+(\s*(\/|of|-)\s*\d+)?$/i;

function fingerprint(text: string): string {
  return text.toLowerCase().replace(DIGITS, "#").trim();
}

export function findFurniture(pages: DocumentPage[]): Set<string> {
  if (pages.length < MIN_PAGES_TO_JUDGE) return new Set();

  const counts = new Map<string, number>();
  for (const page of pages) {
    const edge = [...page.lines.slice(0, EDGE_LINES), ...page.lines.slice(-EDGE_LINES)];
    for (const print of new Set(edge.map((line) => fingerprint(line.text)))) {
      counts.set(print, (counts.get(print) ?? 0) + 1);
    }
  }

  const threshold = pages.length * REPEAT_RATIO;
  return new Set(
    [...counts].filter(([print, n]) => print && n >= threshold).map(([print]) => print),
  );
}

export function stripFurniture(pages: DocumentPage[]): DocumentPage[] {
  const furniture = findFurniture(pages);
  const judgeable = pages.length >= MIN_PAGES_TO_JUDGE;

  return pages.map((page) => {
    const last = page.lines.length - 1;
    const kept = page.lines.filter((line, index) => {
      if (furniture.has(fingerprint(line.text))) return false;
      const atEdge = index < EDGE_LINES || index > last - EDGE_LINES;
      if (atEdge && judgeable && PAGE_NUMBER.test(line.text.trim())) return false;
      return true;
    });
    return {
      text: kept.map((line) => line.text).join("\n"),
      pageNumber: page.pageNumber,
      source: page.source,
      lines: kept,
    };
  });
}
